### ISSUE TITLE
Multifile variety (errors included on purpose)

### ISSUE THREAD

**Issue by flexoron** [infradig/trealla#392]
$ cat a.pl
a :- d.
c :- write(c).

$ cat b.pl
a :- c.
d :- write(d).
x.. % the syntax error rejects file b.pl (atomic transaction)

```
$ trealla/tpl a b
Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.23
?- listing.
a :- d.
c :- write(c).
   true.
?- a.
uncaught exception: error(existence_error(procedure,d/0),d/0)  % d/0 in file b.pl which has been rejected
?- 
```
In contrast
```
?- [user].
a.
   true.
?- listing.
a.
   true.
?- [user].
b..
Error: syntax error, near '..', operator expected 'b..
'
   true.
?- listing. % kind of atomic reflection?
   true.
?-
```

Deduction: Loading a bunch of files should be an atomic transaction, too. (aka all or nothing)

**Comment by flexoron** [comment]
Here another reason why it should be atomic:
```
$ trealla/tpl a b
Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.23
?- listing.
a :- d.
c :- write(c).
   true.

?- [user]. add missing d/0
d :- write(d).
   true.

?- a.
d   true   % unexpected. system waits here. why?
;  false.
?-
?- listing.
a :- d.
c :- write(c).
d :- write(d).
   true.
?-
```

In Contrast: loading error-free that is without b.pl
```
$ trealla/tpl a

?- [user]. % add missing d/0
d :- write(d).
   true.

?- a.
d   true.  % expected
?-
?- listing.
a :- d.
c :- write(c).
d :- write(d).
   true.
?- 
```

**Comment by infradig** [comment]
I don't see consult as being backtrackable, so loading multiple files can't
be atomic.

**Comment by flexoron** [comment]
ok
The second case attracts much more interest
?- a.
d   true   % unexpected. system waits here. why? see above

**Comment by infradig** [comment]
If it's waiting there's a choice-point. The question is why.

**Comment by flexoron** [comment]
yes
also when loading a third file
$ cat c.pl
d :- write(c_d).

```
$ tpl a b c
Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.23
?- a.
c_d   true
```

**Comment by infradig** [comment]
I think loading should *stop* on an error in these cases.

**Comment by flexoron** [comment]
compromise-oriented programming
a touch of atomic
I agree

**Comment by flexoron** [comment]
tokenize is too hard to understand.
the syntax of rule d. is ok
flags,infos about d. are noted in structures
is it true that tokenize does this?

**Comment by flexoron** [comment]
tokenize: fprintf(stderr, "Debug:
thats too much for me

**Comment by flexoron** [comment]
Now this is really strange

Both increments not removed: dbe->cl.ugen_erased = ++m->pl->ugen;
```
$ tpl a b
Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.23-dirty
?- [user].
d.
   true.
?- a.
   true  % choice point

But if you call a. before adding d. no wait

Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.23-dirty
?- a.
uncaught exception: error(existence_error(procedure,d/0),d/0)
?- [user].
d.
   true.
?- a.
   true. % no choice point
?-
```

**Comment by infradig** [comment]
Not updating the generation just makes retracted clauses visible. Are you
saying the predicate should be abolished? I suppose it should if it created
it. I'd need to check.

**Comment by flexoron** [comment]
The trouble is the predicate name a/0.
Solution: don't allow single char ``a`` to be a predicate name. :-)

$ cat a.pl
x :- d.
a :- d.
aa :- d.

```
Trealla Prolog (c) Infradig 2020-2021, v1.19.23
?- [user].
d.
   true.
?- aa.
   true.
?- x.
   true.
?- a.
   true
;  ... .
```

**Comment by infradig** [comment]
You've lost me on this one.

**Comment by infradig** [comment]
Try with v1.19.27

**Comment by flexoron** [comment]
```
$ tpl a b
Error: syntax error, near '..', operator expected 'x.. % the syntax-error
'
Trealla Prolog (c) Infradig 2020-2021, v1.19.27
?- [user].
d.
   true.
?- aa.
   true.
?- x.
   true.
?- a.
   true
;  ... .
?-
```

**Comment by infradig** [comment]
What's in b.pl ?

**Comment by flexoron** [comment]
every multifile predicate has this issue.
```
$ cat a.pl
x :- d.
a :- d.
aa :- d.
c :- write(c).

$ cat b.pl
a :- c.
x :- c.
d :- write(d).
x.. % the syntax-error
```

**Comment by infradig** [comment]
Ok, I'll look over the weekend.

**Comment by UWN** [comment]
Multifile predicates require 7.4.2.2 multifile/1. And if they do not have this directive, then a warning would be nice. But that's all. Same also for errors during preparation for execution.

**Comment by flexoron** [comment]
fixed with v1.20.7
